### PR TITLE
Yield Execution

### PR DESCRIPTION
**Pull request by saketh-are**
Implements the `promise_yield_create` and `promise_yield_resume` host functions described in [NEP 519](https://github.com/near/NEPs/pull/519).

- `promise_yield_create`: Creates a promise callback with a special input data dependency to be delivered via a resume call.
- `promise_yield_resume`: Delivers input data for a yielded promise, triggering execution of the callback.

A fixed-length timeout `yield_timeout_length_in_blocks` is configured at the protocol level. If a yielded promise is not resumed within the timeout length, the callback is triggered automatically.

Co-authored by: Simonas Kazlauskas <git@kazlauskas.me>

### CODE REVIEW

File: runtime/near-vm-runner/src/logic/logic.rs [hunk 1]
```diff
@@ -2015,6 +2016,152 @@ impl<'a> VMLogic<'a> {
...
+    ///
+    /// The following fees are charged:
+    ///
+    /// * `base` fee;
+    /// * `yield_create_base` fee;
+    /// * `yield_create_byte` for each byte of `method_name` and `arguments`;
+    /// * Fees for reading the `method_name` and `arguments`;
+    /// * Fees for writing the Data ID to the output register;
+    /// * Fees for setting up the receipt.
+    pub fn promise_yield_create(
```
Comment (saketh-are) [inline runtime/near-vm-runner/src/logic/logic.rs:2047]: @nagisa Should we charge for creation of a data receipt as part of `promise_yield_create`, seeing as one will be created by the protocol if the yield ends up timing out? If so, should we also remove that cost from `promise_yield_resume` given that it was prepaid in the yield create call?

File: runtime/runtime/src/actions.rs [hunk 2]
```diff
@@ -307,6 +308,39 @@ pub(crate) fn action_function_call(
...
+            get_yielded_promise_indices(state_update).unwrap_or_default();
+        let initial_yielded_promise_indices = yielded_promise_indices.clone();
+
+        for (account_id, yielded_data_id) in receipt_manager.yielded_data_ids.iter() {
+            set_yielded_promise(
+                state_update,
+                &mut yielded_promise_indices,
+                account_id.clone(),
+                yielded_data_id.clone(),
+                apply_state.block_height + 30, // TODO: properly configure this
```
Comment (saketh-are) [inline runtime/runtime/src/actions.rs:335]: @nagisa Where do you think would be the best place to define this constant?

File: runtime/near-vm-runner/src/logic/logic.rs (hunk 1 above)
Comment (nagisa) [reply runtime/near-vm-runner/src/logic/logic.rs:2047]: Thanks for pointing this out. Indeed, `yield_create_base` fee should cover any additional fees for actions that might occur implicitly after this function has finished executing. In practice what should happen is that the fee reflects the most expensive case of `promise_yield_create` use. I was anticipating of just running the estimator with a contract that never resumes its yields as a point of comparison, but you make a good point that `resume` should not charge at least the base fee for data receipt creation again.

File: runtime/runtime/src/lib.rs [hunk 3]
```diff
@@ -1499,10 +1500,70 @@ impl Runtime {
             prefetcher.clear();
         }
 
+        // Resolve timed-out yielded promises
```
Comment (nagisa) [inline runtime/runtime/src/lib.rs:1503]: style: should perhaps consider splitting this out into a function. The `apply` is quite a bit too large as is.

**Review by nagisa — APPROVED** [review]

**Comment by nagisa** [comment]
On coverage: we're hoping to finish up the PR as a follow up, including test coverage. This PR has been becoming humongous and its less trivial than we'd like to work on our respective parts without this being in mainline.

File: core/parameters/res/runtime_configs/139.yaml [hunk 4]
```diff
@@ -0,0 +1,6 @@
+yield_resume: { old: false, new: true }
+# FIXME(yield_resume): These fees are placeholders.
+wasm_yield_create_base: { old: 300_000_000_000_000, new: 50_000_000_000_000 }
+wasm_yield_create_byte: { old: 300_000_000_000_000, new: 10_000_000 }
+wasm_yield_resume_base: { old: 300_000_000_000_000, new: 50_000_000_000_000 }
+wasm_yield_resume_byte: { old: 300_000_000_000_000, new: 100_000_000 }
```
[the remark this answers is not in the document]
Comment (saketh-are) [reply core/parameters/res/runtime_configs/139.yaml:6]: Yes, I believe they are too high. To allow local testing I have been modifying this file and using much lower costs; it is not really a blocker if they continue to be set too high until we have the proper values.

File: runtime/runtime/src/lib.rs (hunk 3 above)
Comment (saketh-are) [reply runtime/runtime/src/lib.rs:1503]: Makes sense, I will look into sending a follow-up PR refactoring `apply`.

File: core/store/src/lib.rs [hunk 5]
```diff
@@ -751,6 +754,51 @@ pub fn set_delayed_receipt(
...
+pub fn get_yielded_promise(
+    trie: &dyn TrieAccess,
+    data_id: CryptoHash,
+) -> Result<Option<YieldedPromise>, StorageError> {
+    get(trie, &TrieKey::YieldedPromise { data_id })
+}
+
+pub fn remove_yielded_promise(state_update: &mut TrieUpdate, data_id: CryptoHash) {
+    state_update.remove(TrieKey::YieldedPromise { data_id })
+}
```
[the remark this answers is not in the document]
Comment (saketh-are) [reply core/store/src/lib.rs]: The gated compilation turned out to be quite a long thread to pull on (needing to gate any dependent code, associated imports, etc.) so I went with some asserts for now.

**Pull request merged**